proposals/0000-ghc-extensions.rst

```diff
@@ -0,0 +1,249 @@
+``default-language`` field).
+
+When ``ghc`` is used without an explicit language choice, the latest
+``GHC20xx`` known to GHC is used. This applies in particular to uses of
+``ghci``.
+
+
+Criteria
+^^^^^^^^
+
+We define a set of criteria that an extension should guide the committee
+members in their process of picking suitable extensions. These are not hard
+criteria, but often subjective, and not all need to be satisfied in all cases.
+
+1. The extension is (mostly) *conservative*: All programs that were accepted
+   without the extension remain accepted, and with the same meaning. We prefer
+   to take this as a strict criteria, and grant exceptions on a case-by-case
+   basis.
+2. *New failure modes* that become possible with the extension are *rare*
+   and/or easy to diagnose. These failure modes include new error messages,
+   wrong inferred types, and runtime errors, for example. In other words, the
+   *developer experience* should not get worse by adding this extension to the
+   default mix.
+3. The extensions *complement* the design of standard Haskell. For example,
+   ``MultiParamTypeClasses`` extends what's already there, and works well with
+   the rest of features. If we suddenly decided you can use Lisp-like syntax,
+   that would not complement the current design.
+4. The extension has been – and can reasonably be predicted to remain –
+   *stable*.
+5. The extension is one that users might plausibly want to be on all the time.
+   This excludes experimental extensions that deliberately enable
+   potentially-unsafe or unstable features, such as ``IncoherentInstances`` or
+   ``MagicHash``.
+6. The extension has *widespread* usage.
+7. The extension is favored by the community, with many in favor, and very few
+   opposed to its inclusion.
+
+Process
+^^^^^^^
+
+* 4 months before the expected GHC spring release day of 20xx, the committee
+  Secretary starts the GHC20xx process.
+
+  They inform the committee, in an email to the mailing list, of all language
+  extensions supported by the latest released GHC that are not in GHC20(xx-1),
+  which could be added. They also list all extensions *in* GHC20(xx-1), which
+  might be omitted in GHC202(xx-1) (likely a rare thing).
+
+* In order to gather data on the criterium “widespread usage”, the secretary
+  creates a tally of which extensions are used how often on Hackage.
+
+* In order to gather data on the criterium “community support”, the secretary
+  runs a public poll on a suitable platform for one week where anyone can vote
+  in favor or against the inclusion of a given extension, or points the
+  committee to a suitable existing survey result.
+
+**Alternative 1:**
+
+* At the start of the process, the secretary creates a PR with a proposal saying (roughly)
+   
+    GHC20xx contains the following extensions in addition to those in
+    GHC20(xx-1):
+    
+    * (none yet)
+    
+    and removes these extensions
+   
+    * (none yet)
+
```

Comment:
```suggestion

    This PR is a suitable place, besides the poll, for the wider commity to weigh in.
```
@simonpj  like this?

Reply:
```suggestion

    This PR is a suitable place, besides the poll, for the wider
     community to weigh in. The community is invited to follow the
     committee votes, rationales and discussion on the public email
     archive, and if the committee is missing an important piece of
     information (e.g. more code breaking than expected), to raise
     such a point.
 
     We hope, however, that the community poll is sufficient to convey
     the level of community support and demand for specific extensions
     have, and want to discourage lengthy, opinion-based discussions of
     the merits of extensions.
 
     If you miss your favorite extension in the list, please remember
     that you can still use it (by setting the flag explicitly), and
     that it can still go in next round.
```